### ISSUE TITLE
Add new GUID (EFI_BOOT_MANAGER_POLICY_USB_GUID) to EFI_BOOT_MANAGER_POLICY_PROTOCOL (Bugzilla Bug 3347)

### ISSUE THREAD

**Issue by tianocore-issues** [tianocore/edk2#8825]
This issue was created automatically with [bugzilla2github](https://github.com/makubacki/bugzilla2github/tree/bz2gh_tianocore)
# Bugzilla Bug 3347
Date: 2021-04-20T23:40:18+00:00
From: Sunny Wang &lt;<Sunny.Wang>&gt;
To:   unassigned &lt;<unassigned>&gt;
CC:   @ardbiesheuvel, @lersek, pete, samer.el-haj-mahmoud

Last updated: 2021-04-21T10:56:35+00:00

**Comment by tianocore-issues** [comment]
## Comment 15747
Date: 2021-04-20 23:40:18 +0000
From: Sunny Wang &lt;<Sunny.Wang>&gt;

- Industry Specification: UEFI Specification
- Target OS: ---
- Bugzilla Assignee(s): unassigned &lt;<unassigned>&gt;

The new GUID will be used by platform firmware or application to discover USB devices in the Fast Boot path for system diagnostic testing or deployment. We saw this need on ARM systems.

For more information about Fast Boot, please check https://github.com/tianocore/edk2-platforms/commit/efdc159ef7c9f15581a0f63d755a1530ff475156#diff-0975811f2220366c94305668f68704bc5bca2b6918fcfa6f4c1a52a415355011.   

Currently, the classes supported by EFI_BOOT_MANAGER_PROTOCOL.ConnectDeviceClass() are the following three GUIDs for different connecting target devices. We want to add one more GUID to connect USB devices. 
 - EFI_BOOT_MANAGER_POLICY_CONSOLE_GUID 
 - EFI_BOOT_MANAGER_POLICY_NETWORK_GUID
 - EFI_BOOT_MANAGER_POLICY_CONNECT_ALL_GUID

**Comment by tianocore-issues** [comment]
## Comment 15752
Date: 2021-04-21 07:53:39 +0000
From: @lersek

Hi Sunny,

(1) You are referring to EFI_BOOT_MANAGER_PROTOCOL as an existent thing (or maybe a thing already under standardization / code first). Where does it live? I can&apos;t see any trace of it in edk2.

Ah, do you mean EFI_BOOT_MANAGER_POLICY_PROTOCOL?

(2) Regarding the commit you reference (the introduction of &quot;gRaspberryPiTokenSpaceGuid.PcdBootPolicy&quot;) -- is that PCD not redundant with the PI-specified &quot;boot mode&quot; concept?

The boot mode is maintained in the EFI_HOB_HANDOFF_INFO_TABLE (PHIT) HOB, can be managed via the GetBootMode() / SetBootMode() PEI services, and once the boot mode has been determined / finalized, the EFI_PEI_MASTER_BOOT_MODE_PPI (with null interface struct) must be installed. The standard boot modes include BOOT_WITH_FULL_CONFIGURATION, BOOT_WITH_MINIMAL_CONFIGURATION, BOOT_ASSUMING_NO_CONFIGURATION_CHANGES, BOOT_WITH_DEFAULT_SETTINGS, among others. There is also room for reserved encodings. Could one of those perhaps map to your &quot;fast boot&quot; policy?

Asking because &quot;ConnectAll() or not&quot; seems to map to &quot;full config vs. minimal&quot; or something like that (cf. &quot;only required devices will be discovered&quot;, in the UNI file in the commit).

Also, if you want to control the boot mode via the setup browser / HII, that&apos;s possible -- a DXE driver can manage a non-volatile UEFI variable via HII indeed, and in the PEI phase (after reboot), you can have read-only UEFI variable access. You could have a platform PEIM that sets the boot mode (see above) from the UEFI variable.

Note that Platform BDS can also access the boot mode, see GetBootModeHob() [MdePkg/Library/DxeHobLib/HobLib.c].

Basically I would recommend replacing PcdBootPolicy with BOOT_WITH_FULL_CONFIGURATION / BOOT_WITH_MINIMAL_CONFIGURATION.

**Comment by tianocore-issues** [comment]
## Comment 15757
Date: 2021-04-21 10:54:15 +0000
From: Sunny Wang &lt;<Sunny.Wang>&gt;

I just got an internal comment that this proposal may not work on some ARM systems&apos; USB controllers that are not on a common PCI bus. Therefore, let&apos;s close this bug. 

Thanks for the comments, Laszlo. I will still take your valuable comments back to our internal discussion.

**Comment by tianocore-issues** [comment]
## Comment 15758
Date: 2021-04-21 10:56:35 +0000
From: Sunny Wang &lt;<Sunny.Wang>&gt;

I just got an internal comment that this proposal may not work on some ARM systems&apos; USB controllers that are not on a common PCI bus. Therefore, let&apos;s close this bug.
